frontends/concrete-python/examples/levenshtein_distance/levenshtein_distance.py

```diff
@@ -0,0 +1,412 @@
+
+    def set_alphabet(self, alphabet, verbose=True):
+        """Set the alphabet."""
+        self.check_alphabet(alphabet)
+
+        if alphabet == "string":
+            self.set_lowercase()
+        if alphabet == "STRING":
+            self.set_uppercase()
+        if alphabet == "StRiNg":
+            self.set_anycase()
+        if alphabet == "ACTG":
+            self.set_dna()
+
+        if verbose:
+            print(f"Making random tests with alphabet {alphabet}")
+            print(f"Letters are {self.letters}\n")
+
+        for i, c in enumerate(self.letters):
+            self.mapping_to_int[c] = i
+
+    def check_string_is_in_alphabet(self, string):
+        """Check a string is a valid string of an alphabet."""
+        assert len(self.mapping_to_int) > 0, "Mapping not defined"
+
+        for c in string:
+            if c not in self.mapping_to_int:
+                raise ValueError(
+                    f"Char {c} of {string} is not in alphabet {list(self.mapping_to_int.keys())}, please choose the right --alphabet"
+                )
+
+    def _random_pick_in_values(self):
+        """Pick the integer-encoding of a random char in an alphabet."""
+        return numpy.random.randint(len(self.mapping_to_int))
+
+    def _random_pick_in_keys(self):
+        """Pick a random char in an alphabet."""
+        return random.choice(list(self.mapping_to_int))
+
+    def _random_string(self, l):
+        """Pick a random string in the alphabet."""
+        return "".join([self._random_pick_in_keys() for _ in range(l)])
+
+    def prepare_random_patterns(self, len_min, len_max, nb_strings):
+        """Prepare random patterns of different lengths."""
+        assert len(self.mapping_to_int) > 0, "Mapping not defined"
+
+        list_patterns = []
+        for _ in range(nb_strings):
+            for length_1 in range(len_min, len_max + 1):
+                for length_2 in range(len_min, len_max + 1):
+                    list_patterns += [
+                        (
+                            self._random_string(length_1),
+                            self._random_string(length_2),
+                        )
+                        for _ in range(1)
+                    ]
+
+        return list_patterns
+
+    def encode_string(self, string):
+        """Encode a string, ie map it to integers using the alphabet."""
+        return tuple([self.mapping_to_int[si] for si in string])
+
+    def encode_and_encrypt_strings(self, a, b):
+        """Encode a string, ie map it to integers using the alphabet, and then encrypt the integers."""
+        a_as_int = self.encode_string(a)
+        b_as_int = self.encode_string(b)
+
+        a_enc = tuple(self.my_module.equal.encrypt(ai, None)[0] for ai in a_as_int)
+        b_enc = tuple(self.my_module.equal.encrypt(None, bi)[1] for bi in b_as_int)
+
+        return a_enc, b_enc
+
+    def compile_module(self, args):
```

Comment:
we may want to let `compile_module`, `compute_in_simulation` and `compute_in_fhe` outside of the Alphabet. They would take the Alphabet as an input, since it's use for encoding. Tell me what you want @umut-sahin , I am anyway not a big fan of these constructions

Reply:
Oh it should definitely outside. Imagine `Alphabet` is just a helper class to encode/decode strings.

Let's move the compilation outside, maybe even it's own class to contain the alphabet the circuit is compiled with.

```python
class LevenshteinDistance:
    alphabet: Alphabet
    module: fhe.Module

    def __init__(self, alphabet: Alphabet):
        self.alphabet = alphabet

        # generate random inputset, compile the module and keygen

    ...

    def calculate(self, a: str, b: str) -> int:
        self.alphabet.check(a)
        self.alphabet.check(b)

        a_encoded = self.alphabet.encode(a)
        b_encoded = self.alphabet.encode(b)

        # execute the circuit
        # this method can return the timings as well

    ...
```

Reply:
Just realized you can do the check in `encode` method as well!

Reply:
ok now?